server/httpin/gateway/gateway.go

```diff
@@ -3,50 +3,42 @@ package gateway
 import (
 	"net/http"
 
-	"github.com/gorilla/mux"
-
-	"github.com/mattermost/mattermost-plugin-apps/apps"
 	"github.com/mattermost/mattermost-plugin-apps/apps/path"
-	"github.com/mattermost/mattermost-plugin-apps/server/appservices"
-	"github.com/mattermost/mattermost-plugin-apps/server/config"
-	"github.com/mattermost/mattermost-plugin-apps/server/httpin"
-	"github.com/mattermost/mattermost-plugin-apps/server/proxy"
+	"github.com/mattermost/mattermost-plugin-apps/server/httpin/handler"
 )
 
 type gateway struct {
-	conf  config.Service
-	proxy proxy.Service
+	*handler.Handler
 }
 
-func Init(h *httpin.Handler, config config.Service, p proxy.Service, _ appservices.Service) {
-	g := &gateway{
-		conf:  config,
-		proxy: p,
+func Init(h *handler.Handler) {
+	g := gateway{
+		Handler: h.PathPrefix(path.Apps),
 	}
 
-	h = h.PathPrefix(path.Apps)
-
-	// Static
-	h.HandleFunc("/{appid}/"+path.StaticFolder+"/{name}",
-		g.static, httpin.RequireUser).Methods(http.MethodGet)
-
-	// Incoming remote webhooks
-	h.HandleFunc("/{appid}"+path.Webhook,
-		g.handleWebhook).Methods(http.MethodPost)
-	h.HandleFunc("/{appid}"+path.Webhook+"/{path}",
-		g.handleWebhook).Methods(http.MethodPost)
-
-	// Remote OAuth2
-	h.HandleFunc("/{appid}"+path.RemoteOAuth2Connect,
-		g.remoteOAuth2Connect, httpin.RequireUser).Methods(http.MethodGet)
-	h.HandleFunc("/{appid}"+path.RemoteOAuth2Complete,
-		g.remoteOAuth2Complete, httpin.RequireUser).Methods(http.MethodGet)
-}
-
-func appIDVar(req *http.Request) apps.AppID {
-	s, ok := mux.Vars(req)["appid"]
-	if ok {
-		return apps.AppID(s)
-	}
-	return ""
+	// Static: /{appid}/static/...
+	h.HandleFunc(handler.AppIDPath+path.StaticFolder+"/{name}", g.static,
```

Comment:
add `func (h *Handler) HandleAppFunc` that would prepend `/{appid}` and add h.RequireDestinationAppFromPath